src/mod_cache_users.erl

```diff
@@ -0,0 +1,262 @@
+%% Helpers
+%%====================================================================
+
+-spec does_cached_user_exist(Status :: boolean(),
+                             HostType :: mongooseim:host_type(),
+                             Jid :: jid:jid(),
+                             RequestType :: ejabberd_auth:exist_type()) ->
+    boolean().
+does_cached_user_exist(false, HostType, #jid{luser = LUser, lserver = LServer}, stored) ->
+    Key = key(LUser, LServer),
+    ParentTab = tbl_name(HostType),
+    case is_member(tbl_name(HostType), Key, ets:last(ParentTab)) of
+        true -> {stop, true};
+        false -> false
+    end;
+does_cached_user_exist(Status, _, _, _) ->
+    Status.
+
+-spec maybe_put_user_into_cache(Status :: boolean(),
+                                HostType :: mongooseim:host_type(),
+                                Jid :: jid:jid(),
+                                RequestType :: ejabberd_auth:exist_type()) ->
+    boolean().
+maybe_put_user_into_cache(true, HostType, #jid{luser = LUser, lserver = LServer}, stored) ->
+    Key = key(LUser, LServer),
+    ParentTab = tbl_name(HostType),
+    put_member(ParentTab, Key, ets:last(ParentTab));
+maybe_put_user_into_cache(Status, _, _, _) ->
+    Status.
+
+-spec key(jid:luser(), jid:lserver()) -> {jid:lserver(), jid:luser()}.
+key(LUser, LServer) ->
+    {LServer, LUser}.
+
+is_member(_, _, '$end_of_table') ->
+    false;
+is_member(ParentTab, Key, SegmentKey) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            case ets:member(EtsSegment, Key) of
+                false -> is_member(ParentTab, Key, ets:prev(ParentTab, SegmentKey));
+                true -> true
+            end;
+        [] ->
+           is_member(ParentTab, Key, ets:prev(ParentTab, SegmentKey))
+    end.
+
+%% There's a chance that by the time we insert in the ets table, this table is not
+%% the first anymore because the cleaner has taken action and pushed it behind.
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(ParentTab, Key, SegmentKey) ->
+    EtsSegment = ets:lookup_element(ParentTab, SegmentKey, 2),
+    ets:insert(EtsSegment, {Key}).
+
+-spec start_server(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_server(HostType, Opts) ->
+    ParentTab = tbl_name(HostType),
+    Spec = #{id => ParentTab, start => {?MODULE, start_link, [ParentTab, Opts]},
+             restart => permanent, shutdown => 5000,
+             type => worker, modules => [?MODULE]},
+    {ok, _} = ejabberd_sup:start_child(Spec).
+
+-spec stop_server(mongooseim:host_type()) -> any().
+stop_server(HostType) ->
+    ok = ejabberd_sup:stop_child(tbl_name(HostType)).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-record(cache_state, {tab :: atom(),
+                      ttl = timer:hours(8) :: timeout()}).
```

Comment:
It looks like a duplicated default, do we need to have the "8" in two places? They can get out of sync.